src/screen.h

```diff
@@ -192,4 +194,60 @@ bool screen_force_clear_to_end();
 /// Returns the length of an escape code. Exposed for testing purposes only.
 size_t escape_code_length(const wchar_t *code);
 
+// Maintain a mapping of escape sequences to their length for fast lookup.
+class cached_esc_sequences_t {
+   private:
+    // Cached escape sequences we've already detected in the prompt and similar strings.
+    std::set<wcstring> cache;
+    // The escape sequence lengths we've cached. My original implementation used min and max
+    // length variables. The cache was then iterated over using a loop like this:
+    // `for (size_t l = min; l <= max; l++)`.
+    //
+    // However that is inefficient when there are big gaps in the lengths. This has been observed
+    // with the BobTheFish theme which has a bunch of 5 and 6 char sequences and 16 to 19 char
+    // sequences and almost nothing in between. So instead we keep track of only those escape
+    // sequence lengths we've actually cached to avoid checking for matches of lengths we know are
+    // not in our cache.
+    std::set<size_t> lengths;
+
+   public:
+    explicit cached_esc_sequences_t() : cache(), lengths() {}
+
+    void add_entry(const wchar_t *entry, size_t len) {
+        auto str = wcstring(entry, len);
+
+#if 0
+        // This is a can't happen scenario. I only wrote this to validate during testing that it
+        // wouldn't be triggered. I'm leaving it in but commented out in case someone feels the need
+        // to re-enable the check.
+        auto match = cache.find(str);
+        if (match != cache.end()) {
+            debug(0, "unexpected add_entry() call of a value already in the cache: '%ls'",
+                  escape(str.c_str(), ESCAPE_ALL).c_str());
+            return;
+        }
+#endif
+
+        cache.emplace(str);
```

Comment:
`emplace` is nice to avoid the intermediate allocation of `str`. It accepts the same parameters as the constructor for the contained type (wcstring). So `cache.emplace(entry, len)`.

Alternatively `cache.insert(str)` is fine too.